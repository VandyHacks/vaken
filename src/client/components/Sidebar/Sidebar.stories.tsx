import React from 'react';
import { Meta, Story } from '@storybook/react'; // eslint-disable-line import/no-extraneous-dependencies
import { MemoryRouter } from 'react-router-dom';
import Component, { Props } from './Sidebar';
import { GlobalStyle } from '../../app';

export default {
	title: 'Components/Sidebar/Sidebar',
	component: Component,
} as Meta;

export const Sidebar: Story<Props> = args => (
	<MemoryRouter>
		{/* Proper rendering of ul elements relies on the global stylesheet */}
		<GlobalStyle />
		<Component {...args} />
	</MemoryRouter>
);
