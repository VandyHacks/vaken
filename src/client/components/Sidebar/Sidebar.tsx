import React, { useContext } from 'react';
import styled from 'styled-components';
import { NavLink as UglyNavLink, withRouter } from 'react-router-dom';
import SqLogo from '../../assets/img/square_hackathon_logo.svg';
import STRINGS from '../../assets/strings.json';
import NavButton from '../Buttons/NavButton';
import { SpaceBetweenColumn, FlexEndColumn } from '../Containers/FlexContainers';
import { SmallCenteredText } from '../Text/SmallCenteredText';
import routes from '../../assets/routes';
import { AuthContext } from '../../contexts/AuthContext';

const Layout = styled.div`
	grid-area: sidebar;
`;

const Background = styled.div`
	background: linear-gradient(254.59deg, #bd7ae3 0%, #8461c9 100%);
	width: 100%;
	height: 100vh;

	display: flex;
	flex-flow: column nowrap;
	justify-content: flex-start;
	align-items: flex-start;
`;

const Logo = styled.div`
	margin: 3rem 2rem;
	width: min-content;
	align-self: center;
`;

const HorizontalLine = styled.hr`
	margin: 0 2rem;
	width: calc(100% - 4rem);
	border: 0.03125rem solid white;
	margin-bottom: 2rem;
`;

const HorizontalLineLogout = styled(HorizontalLine)`
	margin-top: 0;
	margin-bottom: 1rem;
`;

const NavButtonWhiteText = styled(NavButton)`
	color: white;
	font-weight: 100;
	font-size: 28pt;
	padding-left: 1.33rem;
`;

const ALink = styled.a`
	text-decoration: none;
	text-align: left;
	height: max-content;
	width: 100%;
	outline: 0 !important;

	&:focus,
	&:hover,
	&:visited,
	&:link,
	&:active {
		text-decoration: none;
		outline: 0 !important;
	}

	button:active,
	button:focus {
		outline: 0 !important;
	}

	button::-moz-focus-inner {
		border: 0;
	}
`;

const NavLink = ALink.withComponent(UglyNavLink);

interface Route {
	authLevel: string[];
	path: string;
	displayText: string;
}

const ColumnWithSeparators = styled.ul`
	width: 100%;

	li:not(:first-child):before {
		content: '';
		opacity: 0.2;
		background-color: white;
		height: 1px;
		width: calc(100% - 4rem);
		margin: 0 2rem;
		display: block;
	}

	.active {
		${NavButtonWhiteText} {
			font-weight: 400;
		}

		button {
			background: rgba(247, 245, 249, 0.1);
		}
	}
`;

const Sidebar = withRouter(
	// Removed props for eslint, no-unused-vars
	(): JSX.Element => {
		const currentUser = useContext(AuthContext);
		return (
			<Layout>
				<Background>
					<Logo>
						<img src={SqLogo} alt="VH graphic" />
					</Logo>
					<HorizontalLine />
					<SpaceBetweenColumn height="calc(100% - calc(8rem + 160px))">
						<ColumnWithSeparators>
							{routes.map((route: Route) => {
								return route.authLevel.includes(currentUser.authLevel) ? (
									<li key={route.path}>
										<NavLink to={route.path} activeClassName="active">
											<NavButtonWhiteText text={route.displayText} />
										</NavLink>
									</li>
								) : null;
							})}
						</ColumnWithSeparators>
						<FlexEndColumn height="min-content" paddingBottom="1rem">
							<ALink href="/api/logout">
								<NavButtonWhiteText text="Logout" />
							</ALink>
							<HorizontalLineLogout />
							<SmallCenteredText>{STRINGS.HACKATHON_TITLE}</SmallCenteredText>
						</FlexEndColumn>
					</SpaceBetweenColumn>
				</Background>
			</Layout>
		);
	}
);

export default Sidebar;
