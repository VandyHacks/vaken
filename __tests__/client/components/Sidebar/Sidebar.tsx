import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

import Sidebar from '../../../../src/client/components/Sidebar/Sidebar';

describe('Test Sidebar', () => {
	it('Sidebar renders correctly', async () => {
		const component = renderer
			.create(
				// see https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/guides/testing.md
				<MemoryRouter>
					<Sidebar />
				</MemoryRouter>
			)
			.toJSON();

		expect(component).toMatchSnapshot();
	});
});
