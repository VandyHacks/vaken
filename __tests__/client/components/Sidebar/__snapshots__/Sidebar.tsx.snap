// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar 1`] = `
<div
  className="sc-fYxtnH jgPElb"
>
  <div
    className="sc-tilXH cItYAS"
  >
    <div
      className="sc-hEsumM hzCnxB"
    >
      <icon-mock />
    </div>
    <hr
      className="sc-ktHwxA bXeZNy"
    />
    <div
      className="sc-bdVaJa sc-bwzfXH sc-bZQynM jxTPlc"
      height="calc(100% - calc(8rem + 160px))"
    >
      <ul
        className="sc-jwKygS ccqAml"
      >
        <li>
          <a
            aria-current={null}
            className="sc-cmTdod cPmHKU"
            href="/dashboard"
            onClick={[Function]}
          >
            <button
              className="sc-gqjmRU sc-feJyhm cUZQkA sc-fjdhpX cVvOdu"
            >
              <div
                className="sc-htoDjs sc-feJyhm cUZQkA sc-dnqmqq graBnM"
              >
                Dashboard
              </div>
            </button>
          </a>
        </li>
        <li>
          <a
            aria-current={null}
            className="sc-cmTdod cPmHKU"
            href="/profile"
            onClick={[Function]}
          >
            <button
              className="sc-gqjmRU sc-feJyhm cUZQkA sc-fjdhpX cVvOdu"
            >
              <div
                className="sc-htoDjs sc-feJyhm cUZQkA sc-dnqmqq graBnM"
              >
                Profile
              </div>
            </button>
          </a>
        </li>
        <li>
          <a
            aria-current={null}
            className="sc-cmTdod cPmHKU"
            href="/application"
            onClick={[Function]}
          >
            <button
              className="sc-gqjmRU sc-feJyhm cUZQkA sc-fjdhpX cVvOdu"
            >
              <div
                className="sc-htoDjs sc-feJyhm cUZQkA sc-dnqmqq graBnM"
              >
                Application
              </div>
            </button>
          </a>
        </li>
        <li>
          <a
            aria-current={null}
            className="sc-cmTdod cPmHKU"
            href="/team"
            onClick={[Function]}
          >
            <button
              className="sc-gqjmRU sc-feJyhm cUZQkA sc-fjdhpX cVvOdu"
            >
              <div
                className="sc-htoDjs sc-feJyhm cUZQkA sc-dnqmqq graBnM"
              >
                Team
              </div>
            </button>
          </a>
        </li>
        <li>
          <a
            aria-current={null}
            className="sc-cmTdod cPmHKU"
            href="/help"
            onClick={[Function]}
          >
            <button
              className="sc-gqjmRU sc-feJyhm cUZQkA sc-fjdhpX cVvOdu"
            >
              <div
                className="sc-htoDjs sc-feJyhm cUZQkA sc-dnqmqq graBnM"
              >
                Help
              </div>
            </button>
          </a>
        </li>
      </ul>
      <div
        className="sc-bdVaJa sc-bwzfXH sc-EHOje jxVFlK"
        height="min-content"
      >
        <a
          className="sc-iELTvK clnfBm"
          href="/api/logout"
        >
          <button
            className="sc-gqjmRU sc-feJyhm cUZQkA sc-fjdhpX cVvOdu"
          >
            <div
              className="sc-htoDjs sc-feJyhm cUZQkA sc-dnqmqq graBnM"
            >
              Logout
            </div>
          </button>
        </a>
        <hr
          className="sc-ktHwxA sc-kafWEX kUvRYH"
        />
        <h1
          className="sc-jzJRlG cRPkSq"
        >
          VandyHacks 2019
        </h1>
      </div>
    </div>
  </div>
</div>
`;
